Sprint 3

antonbezushko opened the pull request:
Основные силы ушли на ~~шаблонизатор~~ (я изначально неправильно понял, что делает шиблонизатор, поэтому реализовано у меня не так, как в обучении, хотя пример с использыванием блока (Block) у меня есть -> `страница /block`)
Особенности:
- Работа с Virtual DOM (обновляются только изменения в реальном DOMe)
- Есть возможность использывать хуки в компонентах (onMounted, onUpdate) - аналог EventBus у Block
- Реактивность (паттерн Observer)
- JXS (реализован с использыванием своей рендер-функции)
По спринту:
- Роутинг реализован тоже с использыванием компонентов, несуществующие роуты перенаправляются на `/404`. Если юзер авторизован перенправляет на `/messenger` и т.д
- Защита от XSS, я думаю, не требуется, т.к не могу смоделить ситуацию, в который может такой тип атаки сработать, т.к с реальным DOMом я не работаю
- Что реализовано по работе с чатом:
-- Загрузка списка чатов (название, аватар пустой (не реализовал загрузку картинки при создании чата), последнее сообщение, кол-во непрочитанных сообщений)
-- Создание/удаление чата
-- Добавление/удаление пользователей из чата
-- Отправка текстовых сообщений в чат (автоматически обновляется последнее сообщение в списке частов)
-- Работа с данным пользователя (выход, вход, изменение данных)
- На что не хватило времени:
-- Не везде крутилки (loader) есть
-- Загрузки списка с offset для сообщений (пока только последние 20 загружаю + все новые в real-time), список чатов (там по умолчанию около 10 последних выводится), 
-- Работа с файлами

Commit:
Добавил реактивность

Commit:
Обновил функцию reactive

Commit:
Заменил все модули

Commit:
Перевод верстки на новую систему

Commit:
Рефакторинг

Commit:
Валидация страницы авторизации

Commit:
Добавил типы файлов для импорта ts

Commit:
Удалил файлы js

Commit:
Добавил линтеры + рефакторинг

Commit:
Добавил описание с main ветки

Commit:
Добавил typescript, установил фикс. версии пакетов

Commit:
Sprint 2 fix

Commit:
Рефакторинг

Commit:
Реализовал Block и EventBus

Commit:
Сделал тестовую страницу с демонстрацией работы Block

Commit:
Фикс реактивности

Commit:
Фикс реактивности

Commit:
Merge branch 'sprint_2' into deploy

Commit:
Фикс реактивности

Commit:
CRLF -> LF

Commit:
Merge branch 'sprint_2' into deploy

Commit:
feature: Добавил работу с условным рендером

Commit:
Добавил конфиг для prettier

Commit:
Вернул обратно старую реализацию реактивности

Commit:
Реализовал Router

Commit:
Адаптировал страницы для работы с роутингом

Commit:
Настроит prettier

Commit:
Добавил input.component value пропс

Commit:
Добавил работу с пользовательскими данными

Commit:
Before tranform to block

Commit:
Подключил работу с сообщениями

Commit:
Добавил в vdom компоненты

Commit:
Добавил хуки жизненного цикла компонента

Commit:
Рефакторинг

Commit:
Добавил вывод сообщения с переносом строки

Commit:
Merge branch 'sprint_3' into deploy

Commit:
Финальные правки

Commit:
Обновил скрипт запуска

Commit:
Обновил скрипт запуска

Commit:
Очистил данные для входа по умолчанию

gennady-bars commented on src/services/chatService.ts:
```diff
@@ -0,0 +1,248 @@
+      this.store.loadingDeleteChat = true
+      await this.api.deleteChat(chatId)
+      this.getChats()
+    } catch (error) {
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars commented on src/api/authApi.ts:
```diff
@@ -0,0 +1,59 @@
+import { IUser } from '../types/user'
+
+const authApiInstance = new HTTPTransport(
+  'https://ya-praktikum.tech/api/v2/auth',
```
`https://ya-praktikum.tech/api/v2` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

А еще лучше поместить этот базовый урл  1 раз внутрь класса `HTTPTransport`, а в конструктор экземпляров передавать только эндпоинт

```jsx

new HTTPTransport('/auth/');

```

gennady-bars commented on src/api/chatApi.ts:
```diff
@@ -0,0 +1,99 @@
+import { IUserWithRole } from '../types/user'
+
+const chatApiInstance = new HTTPTransport(
+  'https://ya-praktikum.tech/api/v2/chats',
```
`https://ya-praktikum.tech/api/v2` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

А еще лучше поместить этот базовый урл  1 раз внутрь класса `HTTPTransport`, а в конструктор экземпляров передавать только эндпоинт

```jsx

new HTTPTransport('/auth/');

```

gennady-bars commented on src/api/userApi.ts:
```diff
@@ -0,0 +1,90 @@
+import { createFormData } from '../utils/index'
+
+const userApiInstance = new HTTPTransport(
+  'https://ya-praktikum.tech/api/v2/user',
```
`https://ya-praktikum.tech/api/v2` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

А еще лучше поместить этот базовый урл  1 раз внутрь класса `HTTPTransport`, а в конструктор экземпляров передавать только эндпоинт

```jsx

new HTTPTransport('/auth/');

```

antonbezushko replied:
Вынес в отдельную переменную

gennady-bars commented on src/consts.ts:
```diff
@@ -0,0 +1 @@
+export const apiResourceUrl = 'https://ya-praktikum.tech/api/v2/resources/'
```
`https://ya-praktikum.tech/api/v2` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

для картинок тоже нужно использовать базовый урл (он тут тоже такой)


```js


export const apiResourceUrl = `${BASE_URL}/resources/`

```

antonbezushko replied:
Вынес в отдельную переменную

gennady-bars commented on src/modules/fetch.ts:
```diff
@@ -0,0 +1,152 @@
+    this.baseUrl = baseUrl
+  }
+
+  get<T>(url: string, options?: HttpTransportOptions) {
```
**Можно лучше**
Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище

```jsx

// создаем тип метода
type HTTPMethod<T> = (url: string, options?: HttpTransportOptions) => Promise<T>

export class HTTPTransport {

  // используем тип и удаляем дублирование в аргументах
  get: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.GET}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  put: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.PUT}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  post: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.POST}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  delete: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.DELETE}, options.timeout)
  )

```

antonbezushko replied:
Не получалось добавить Generic если вывести в отдельный тип

gennady-bars commented on package.json:
```diff
@@ -3,26 +3,41 @@
-    "dev" : "npx parcel serve ./src/pages/**/*.html ./src/pages/index.html --no-cache",
-    "start" : "npx parcel build ./src/pages/**/*.html ./src/pages/index.html && node server.js",
-    "build" : "npx parcel build ./src/pages/**/*.html ./src/pages/index.html"
+    "dev": "npx parcel serve ./src/index.html ./src/index.html --no-cache",
```
**Можно лучше**

Чтобы **автоматически запускался** проект в браузере, нужно дописать в скрипт `--open`

antonbezushko replied:
Добавил

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная



* `Readme` хорошо оформлен
* Отлично, что фиксируете версии зависимостей
* Отлично, что нет `EOF` ошибок в гите
* Хорошая структура папок и файлов
* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

но есть некоторые недочеты: 

* не могу отправить сообщение в чат. Нет кнопки для отправки 
* после удаления чата в консоли ошибка и перестает все работать https://disk.yandex.ru/i/wHkvUfEb0j6BzQ (это в задеплоенной версии не работает)
* если я перезагружаю страницу, находясь в чатах, то попадаю на страницу 404 от Нетлифай, а по заданию я должен попасть туда, где перезагружал (это в задеплоенной версии не работает)
* при запуске проекта в консоли предупреждения про неуникальные `id` https://disk.yandex.ru/i/PbaHioXXF2Gmbg 
* `https://ya-praktikum.tech/api/v2` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.




**Можно лучше**


* Скрин https://disk.yandex.ru/i/KCZzHrxHF5zImg В моих чатах есть **аватар**, но он не отображается сейчас у Вас. Нужно правильно подключать картинки 
* Должна быть возможность менять картинку в чате
* Попробуйте избавиться от типов `any`
* Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище

Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

Commit:
Добавил константу API_BASE_URL

Commit:
Вынес создание данных формы в класс

Commit:
Добавил глобальный класс formControl

Commit:
Обновил работу с чатом

- Добавил кнопку для отправки сообщения
- Возможность добавления авы в чат при создании
- При удалении чата не грузится весь список чатов, и закрывается WS соединение
- При добавлении чата не грузится весь список чатов

Commit:
Добавил возможность обновить аватар чата

Commit:
Рефакторинг

Commit:
Фикс роутинга на деплое

Commit:
Фикс деплоя на netlify

Krezo commented:
> 

Все необходимые изменения внес

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
